### ISSUE TITLE
[Bug] Same dataset reached via two blend paths is loaded twice with identical sample order

### ISSUE THREAD

**Issue by zsombor-kaiko** [NVIDIA/Megatron-Energon#266]
## Summary

When the same leaf dataset appears twice in a nested `MetadatasetV2` blend, Energon does not detect it. The dataset is instantiated twice, each copy gets its own `WebdatasetSampleLoaderDataset`, and because both are seeded from the same `worker_seed()` they emit **the exact same sample sequence**. The model therefore sees the same samples twice before the dataset is exhausted once, without any warning.

This is easy to hit in practice: blend files that include sub-blend files (e.g. a "replay" blend that pulls in a previous stage's blend, plus an explicit re-include of one sub-blend at a higher weight).

Checked against `main` at `ace766d6edcb1e4e7ca53bd18014729e2d112c37` (2026-08-27)

## Where it happens

- `MetadatasetBlend.get_datasets` (`src/megatron/energon/metadataset/metadataset_v2.py`) appends every leaf returned by the inner references to a flat list and scales the weight. Nothing keys on the resolved path, so a duplicated leaf becomes two `LoadedDataset` entries. `Metadataset` (v1) and `traverse()` behave the same.
- `TaskEncoder.build_train_datasets` builds each entry into its own `RepeatDataset` → `WebdatasetSampleLoaderDataset` and blends them in `BlendDataset`.
- `WorkerRng` is seeded from `WorkerConfig.worker_seed()`, which hashes only `(global_worker_id, seed_offset)`. The dataset identity is not part of the seed, so two loaders over the same shards on the same worker produce identical shuffles. The per-entry `worker_rotation_offset` only rotates which workers get the remainder samples.

## Reproduction

One 100-sample webdataset, blended once directly and once via a nested metadataset, one worker.

```python
import tempfile
from collections import Counter
from pathlib import Path

import webdataset as wds

from megatron.energon import WorkerConfig, get_loader, get_train_dataset
from megatron.energon.flavors import BaseWebdatasetFactory
from megatron.energon.flavors.webdataset.config import MAIN_FOLDER_NAME
from megatron.energon.task_encoder.base import DefaultTaskEncoder

N = 100
tmp = Path(tempfile.mkdtemp())
ds = tmp / "ds1"
(ds / "parts").mkdir(parents=True)
with wds.ShardWriter(f"{ds}/parts/data-%d.tar", maxcount=10) as w:
    for k in range(N):
        w.write({"__key__": f"{k:06d}", "txt": f"{k}".encode()})
    total = w.shard
BaseWebdatasetFactory.prepare_dataset(
    ds, [f"parts/data-{{0..{total - 1}}}.tar"], split_parts_ratio=[("train", 1.0)], shuffle_seed=None
)
(ds / MAIN_FOLDER_NAME / "dataset.yaml").write_text(
    "sample_type:\n  __module__: megatron.energon\n  __class__: TextSample\nfield_map:\n  text: txt\n"
)

HEADER = "__module__: megatron.energon\n__class__: MetadatasetV2\nsplits:\n  train:\n    blend:\n"
(tmp / "inner.yaml").write_text(HEADER + "      - weight: 1\n        path: ./ds1\n")
(tmp / "dup.yaml").write_text(
    HEADER + "      - weight: 1\n        path: ./ds1\n      - weight: 1\n        path: ./inner.yaml\n"
)
(tmp / "single.yaml").write_text(HEADER + "      - weight: 1\n        path: ./ds1\n")

wc = WorkerConfig(rank=0, world_size=1, num_workers=0)


def first_epoch_stats(mds: Path) -> tuple[int, int]:
    loader = get_loader(get_train_dataset(
        mds, worker_config=wc, batch_size=1, shuffle_buffer_size=None,
        max_samples_per_sequence=None, task_encoder=DefaultTaskEncoder(),
    ))
    it = iter(loader)
    counts = Counter(next(it).text[0] for _ in range(N))
    return len(counts), sum(1 for v in counts.values() if v > 1)


print("single :", first_epoch_stats(tmp / "single.yaml"))  # (100, 0)
print("dup    :", first_epoch_stats(tmp / "dup.yaml"))     # (51, 49)
```

Output on `main`:

| blend | unique keys in first 100 draws | keys seen twice |
| --- | --- | --- |
| `single.yaml` | 100 | 0 |
| `dup.yaml` | 51 | 49 |

Inspecting the `BlendDataset` index in `__restore_key__` shows the two entries emit the identical sequence (`['50', '90', '91', '20', '70', '80', '0', '1', ...]` for both). The blend drew 51 samples from entry 0 and 49 from entry 1, and every one of the 49 was a sample entry 0 had already emitted. In general the number of repeats is `min(k, N-k)` where `k` is how many draws went to entry 0.

## Expected behaviour options

1. **Raise** at `get_datasets()` / `traverse()` time when the same resolved leaf path (with the same `split_part`, `subset`, and `subflavors`) appears more than once in one split. This is almost always a configuration mistake.
2. **Warn** and **merge** duplicates into a single `LoadedDataset` whose weight is the sum of the individual weights, so only one loader is built. The effective sampling ratio would be unchanged, but the sample-level duplication disappears.

Option 1 is the safer default, but Option 2 is more flexible so I'd advocate for Option 2. If safety is a concern merging could also be dependent on a parameter.

**Comment by sylvesterkaczmarek** [comment]
I will add a regression for the nested-blend reproduction and coalesce identical leaf references by summing their sampling weights. Different splits, subsets, subflavors, loader configurations, and auxiliary data will remain separate; epochized and joined datasets will keep their existing behavior.
